### PR TITLE
#13708 Support undo/redo operations for fk create and delete

### PR DESCRIPTION
**Pull request by E1izabeth**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationCreateCommand.java [hunk 1]
```diff
@@ -167,19 +181,25 @@ protected ERDAssociation createAssociation(ERDElement<?> sourceEntity, ERDElemen
...
                 },
                 srcAttrs,
                 refAttrs);
+            DBVEntityForeignKey vfk = virtualFkSupplier.get();
             if (vfk == null) {
+                this.virtualFk = null;
+                this.virtualFkSupplier = null;
                 return null;
+            } else {
+                vEntity.persistConfiguration();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationCreateCommand.java]: Configuration is already persisting in `ERDEditorEmbedded::doSave`. Not sure if it should be done here too.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationDeleteCommand.java [hunk 2]
```diff
@@ -65,21 +72,20 @@ public AssociationDeleteCommand(AssociationPart part) {
...
+            if (!UIUtils.confirmAction("Delete logical key", "Are you sure you want to delete logical key '" + association.getName() + "'?")) {
                 return;
             }
-            DBVEntity vEntity = DBVUtils.getVirtualEntity(entityAssociation.getParentObject(), false);
             if (vEntity == null) {
                 UIUtils.showMessageBox(UIUtils.getActiveWorkbenchShell(), "No virtual entity", "Can't find association owner virtual entity", SWT.ICON_ERROR);
                 return;
             }
             vEntity.removeForeignKey((DBVEntityForeignKey) entityAssociation);
+            vEntity.persistConfiguration();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationDeleteCommand.java]: Configuration is already persisting in `ERDEditorEmbedded::doSave`. Not sure if it should be done here too.

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/object/struct/EditForeignKeyPage.java [hunk 3]
```diff
@@ -1013,55 +1014,83 @@ private void setReferenceAttributes(Collection<? extends DBSEntityAttribute> ref
...
-            columns.add(
-                new DBVEntityForeignKeyColumn(
-                    virtualFK, tableColumn.getOwnColumn().getName(), tableColumn.getRefColumn().getName()));
-        }
-        virtualFK.setAttributes(columns);
-        vEntity.addForeignKey(virtualFK);
-        return virtualFK;
+        return new Supplier<DBVEntityForeignKey>() {
+            private List<FKColumnInfo> tableColumns = null;
+            private DBSEntityConstraint constraint = null;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/object/struct/EditForeignKeyPage.java]: Why are they fields? They don't seem to be used anywhere but in the `get` method.

File: plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationCreateCommand.java [hunk 4]
```diff
@@ -137,6 +145,12 @@ public void undo() {
         if (association != null) {
             sourceEntity.removeAssociation(association, true);
             targetEntity.removeReferenceAssociation(association, true);
+
+            if (virtualFk != null) { 
+                virtualFk.getEntity().removeForeignKey(virtualFk);
+                virtualFk.getEntity().persistConfiguration();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/command/AssociationCreateCommand.java]: Configuration is already persisting in `ERDEditorEmbedded::doSave`. Not sure if it should be done here too.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/object/struct/EditForeignKeyPage.java (hunk 3 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.ui.editors.base/src/org/jkiss/dbeaver/ui/editors/object/struct/EditForeignKeyPage.java]: These fields store information. They are read and changed between calls.

**Pull request closed without merging**